docs/rfd.md

```diff
@@ -0,0 +1,478 @@
+
+On startup, the server will do the following (simplifying, there will be additional abstractions here):
+
+1. Ensure its euid is 0 so it can manage cgroups, and ensure that cgroups v2 is enabled.
+2. Create a parent cgroup in `/user.slice/user-<uid>.slice/jobserver` if it doesn't exist. The specific path can be determined by first identifying its own cgroup from `/proc/self/cgroup` (which, if run as `sudo jobserver serve`, will be something like `/user.slice/user-<uid>.slice/<terminal emulator's cgroup>/vte-spawn-<some uuid>.scope`), then walking the tree up until it finds a cgroup matching `user-*.slice`.
+3. Enable the `cpu`, `memory`, and `io` controllers in the `jobserver` cgroup's `cgroup.subtree_control` file, if necessary.
+
+The lifecycle of a job's cgroup will be as follows:
+
+1. When a job is created with resource limits, the server will create a new cgroup with the name `<job-id>` under the parent `jobserver` cgroup.
+2. The server will write the requested resource limits to the new cgroup's `cpu.max`, `memory.max`, and `io.max` files (for whichever limits were specified).
+3. The server will create a file descriptor to the new cgroup by opening its path (`/sys/fs/cgroup/user.slice/user-<uid>.slice/jobserver/<job-id>`).
+4. The server will configure the `exec.Cmd`'s `SysProcAttr` with `CgroupFd` set to the file descriptor. When this is set, Go will call `clone3` with `CLONE_INTO_CGROUP` so that the new child process is started directly in the new cgroup.
+5. When the process exits, the server will close the file descriptor and delete the job's cgroup.
+
+Other notes:
+
+- When clients are streaming logs for a running process, the server will close those streams once the process has exited and all logs have been written.
+- The server will not attempt to modify the controllers of its parent cgroups, so if `user-<uid>.slice` doesn't have any of the required controlers enabled, it will simply bail out.
+- The user will be able to override the default parent cgroup path with a flag, since there can be some variation in the cgroup hierarchy depending on the system.
+
+### Security
+
+#### Authentication and Authorization
+
+Authentication is implemented using simple mTLS, with a subject name encoded in the client certificate's common name field.
+
+TLS 1.3 will be enforced by the server. The Go implementation uses a fixed set of 3 very strong cipher suites - `chacha20-poly1305-sha256`, `aes-128-gcm-sha256`, and `aes-256-gcm-sha384` (no aes ccm modes) - and requires no additional configuration or tuning outside of setting the minimum TLS version to 1.3.
+
+To authorize a user for a specific api method, the server will match the client's subject name against the rbac rules it loaded from the config file,
+then check to see if that subject is named in a role binding for which the associated role allows access to the method the client is calling.
```

Comment:
just food for thought (as https://github.com/kralicky/jobserver/pull/1/files#r1425022556 will likely lead to a simplification of this model anyway): is there some other source of trusted information already in play that could be used here to decide which roles any given client is supposed to have, rather than having a dedicated configuration mechanism tied to the server?

Reply:
Do you mean encoding role information into certificates?

Reply:
Yep! 👍 You're delegating authn for the identity to the issuer of the client certs, why not also roles?